io/src/ply/ply_parser.cpp

```diff
@@ -99,465 +97,396 @@ bool pcl::io::ply::ply_parser::parse (const std::string& filename)
-      }
+    // end_header
+    else if (keyword == "end_header")
+    {
+      if (!end_header_callback_ ())
+          return true;
+      break;
+    }
+    // unknown keyword
+    else
+    {
+      warning_callback_ (line_number_, "ignoring line '" + line + "'");
     }
   }
 
   if (number_of_format_statements == 0)
   {
-    if (error_callback_) 
-     error_callback_ (line_number_, "parse error: a format statement is required");
+    error_callback_ (line_number_, "parse error: a format statement is required");
     return false;
   }
 
   // ascii
   if (format == ascii_format)
   {
-    for (auto element_iterator = elements.cbegin ();
-         element_iterator != elements.cend ();
-         ++element_iterator)
+    for (const auto element_ptr: elements)
     {
-      struct element& element = *(element_iterator->get ());
+      auto& element = *(element_ptr.get ());
       for (std::size_t element_index = 0; element_index < element.count; ++element_index)
       {
-        if (element.begin_element_callback) 
+        if (element.begin_element_callback)
           element.begin_element_callback ();
         if (!std::getline (istream, line, line_delim))
         {
-          if (error_callback_)
-            error_callback_ (line_number_, "parse error");
+          error_callback_ (line_number_, "parse error: found less elements than declared in the header");
           return false;
         }
         istream.ignore (char_ignore_count);
         ++line_number_;
         std::istringstream stringstream (line);
         stringstream.unsetf (std::ios_base::skipws);
         stringstream >> std::ws;
-        for (auto property_iterator = element.properties.cbegin ();
-             property_iterator != element.properties.cend ();
-             ++property_iterator)
+
+        for (const auto property_ptr: element.properties)
         {
-          struct property& property = *(property_iterator->get ());
+          auto& property = *(property_ptr.get ());
           if (!property.parse (*this, format, stringstream))
+          {
+            error_callback_ (line_number_, "parse error: element property count doesn't match the declaration in the header");
             return false;
+          }
         }
         if (!stringstream.eof ())
         {
-          if (error_callback_)
-            error_callback_ (line_number_, "parse error");
+          error_callback_ (line_number_, "parse error: element contains more properties than declared");
           return false;
         }
         if (element.end_element_callback)
           element.end_element_callback ();
       }
     }
     istream >> std::ws;
-    if (istream.fail () || !istream.eof () || istream.bad ())
+    if (istream.fail ())
     {
-      if (error_callback_)
-        error_callback_ (line_number_, "parse error");
-      return false;
+      warning_callback_ (line_number_, "parse error: No newline at the end of file");
```

Comment:
0. Are we sure that failbit necessarily means "no newline at end"? Could there be other situations that lead to failbit being set?
1. Probably not "parse error" if we are triggering warning callback
2. `No` -> `no`
3. Do we want to report this at all?

Reply:
* 0. [This](https://en.cppreference.com/w/cpp/io/ios_base/iostate) is comprehensive. It says failbit is set when the operation to get data from input stream fails to extract correctly, ie.: can't read a char, time, number, etc. based on the overload used. Since we tried extracting white-space before, the only way it would fail is if there were no whitespace chars.
* 1 & 2. Made a mistake there, fix incoming
* 3 Yes, since line ending are required for all lines in non-empty files. See [definition of text file](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_392)